File: 1429-first-unique-number/1429. First Unique Number.py
class FirstUnique:

    def __init__(self, nums: List[int]):
        self.uniqueNums = {}
        self.seen = {}
        for n in nums:
            if  n in self.uniqueNums:
                self.uniqueNums.pop(n)
                self.seen[n] = 1
            elif n in self.seen:
                continue
            else:
                self.uniqueNums[n] = 1

        

    def showFirstUnique(self) -> int:
        for key in self.uniqueNums:
            return key
        return -1

    def add(self, value: int) -> None:
        if value in self.uniqueNums:
            self.uniqueNums.pop(value)
            self.seen[value] = 1
        elif value in self.seen:
            return
        else:
            self.uniqueNums[value] = 1
        


# Your FirstUnique object will be instantiated and called as such:
# obj = FirstUnique(nums)
# param_1 = obj.showFirstUnique()
# obj.add(value)

File: 1429-first-unique-number/1429. First Unique Number.cpp
class FirstUnique {
public:
    FirstUnique(vector<int>& nums) {
        n = nums; 
        count = nums.size(); 
        for (int i = 0; i < nums.size(); ++i) { 
            auto it = map.find(nums[i]);
            if (it == map.end()) { 
                map.insert({nums[i], i}); 
            }
            else { 
                it->second = -1; 
            }
        }
    }
    
    int showFirstUnique() {
        for (int i = 0; i < n.size(); ++i) { 
            auto it = map.find(n[i]); 
            if (it != map.end()) {
                if (it->second != -1) { 
                    return n[i];
                }
            }
        }
        return -1;
    }
    
    void add(int value) {
        auto it = map.find(value); 
        if (it != map.end()) {
            it->second = -1;
        }
        else { 
            map.insert({value, static_cast<int>(n.size())});
        }
        n.push_back(value);

    }
private: 
    vector<int> n;
    unordered_map<int, int> map; 
    int count; 
};

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique* obj = new FirstUnique(nums);
 * int param_1 = obj->showFirstUnique();
 * obj->add(value);
 */